index.js

```diff
@@ -1,86 +1,57 @@
+    collection = [collection];
+  }
 
-  // applies the fetch method one time each to an array of items:
-  const applyMultipleMethod = (collectionItem, done) => {
-    const allFetchedItems = [];
-    async.eachSeries(collectionItem[key], (listItem, eachDone) => {
-      if (cache[listItem]) {
-        assign(allFetchedItems, cache[listItem]);
-        return eachDone();
-      }
-      method(listItem, (err, fetchedItem) => {
-        if (err) {
-          return done(err);
-        }
-        cache[listItem] = fetchedItem;
-        assign(allFetchedItems, fetchedItem);
-        eachDone();
+  // get unique list of keys in the collection:
+  const set = new Set();
+  collection.forEach((item) => {
+    if (Array.isArray(item[key])) {
+      item[key].forEach((curKey) => {
+        set.add(curKey);
       });
-    }, (err) => {
-      if (err) {
-        return done(err);
-      }
-      collectionItem[property] = allFetchedItems;
-      return done();
-    });
-  };
-
-  const handleCallback = (err) => {
-    if (err) {
-      return callback(err);
+    } else {
+      set.add(item[key]);
     }
-    return callback(null, collection);
-  };
+  });
+  const uniqueKeys = Array.from(set);
 
-  // if collection was just one item, just process and return it by itself:
-  if (!Array.isArray(collection)) {
-    if (!Array.isArray(collection[key])) {
-      return applySingleMethod(collection, handleCallback);
-    }
-    // the collection's field is an array of items to fetch here:
-    return applyMultipleMethod(collection, handleCallback);
+  // get unique list of items needed by the collection:
+  const promiseList = [];
+  uniqueKeys.forEach((itemKey) => {
+    promiseList.push(method(itemKey));
+  });
+  const uniqueItemList = await Promise.all(promiseList);
+
+  // organize the unique items by key:
+  const uniqueItemDict = {};
+  for (let i = 0; i < uniqueKeys.length; i++) {
+    uniqueItemDict[uniqueKeys[i]] = uniqueItemList[i];
   }
 
-  // otherwise, if collection is multiple items:
-  return async.each(collection, (collectionItem, eachDone) => {
-    // if each field of the collection item is one item:
-    if (!Array.isArray(collectionItem[key])) {
-      return applySingleMethod(collectionItem, eachDone);
+  // now set the requested property field using the fetched items:
+  collection.forEach((item) => {
+    const curKey = item[key];
+    let setProperty = () => {
+      item[property] = get ? uniqueItemDict[curKey][get] : uniqueItemDict[curKey];
+    };
+    if (Array.isArray(curKey)) {
+      setProperty = () => {
+        const allValues = [];
+        curKey.forEach((individualKey) => {
+          allValues.push(get ? uniqueItemDict[individualKey][get] : uniqueItemDict[individualKey]);
+        });
+        item[property] = allValues;
+      };
     }
-    // if the collection item is a list of multiple items:
-    return applyMultipleMethod(collectionItem, eachDone);
-  }, handleCallback);
+    setProperty();
```

Comment:
you aren't really saving any effort in creating a function and then overriding.  probably just better with an if else